geopmctlpy/geopmctl/metric.py

```diff
@@ -0,0 +1,680 @@
+    file. Passed to the Metric instance created from this signal as sample_only.
+    '''
+    return_val = None
+    behavior = Behavior(pio.signal_info(descriptor[0])[2])
+    if behavior is Behavior.MONOTONE:
+        return_val = MonotoneSignal(descriptor, report_formatter=report_formatter,
+                                    trace_formatter=trace_formatter)
+    else:
+        return_val = Signal(descriptor, behavior, report_formatter=report_formatter,
+                            trace_formatter=trace_formatter)
+    return return_val
+
+
+class Behavior(Enum):
+    '''
+    Enum representing Signal behaviors. Should match the integer representation of behaviors
+    returned by pio.
+    '''
+    CONSTANT = 0
+    MONOTONE = 1
+    VARIABLE = 2
+    LABEL = 3
+
+
+class SignalList:
+    '''
+    Class storing a collection of signals. A signal list is a representative of the signals passed
+    to the Agent.update() method.
+    '''
+
+    def __init__(self):
+        self._signals = []
+
+    def update(self, signal_values):
+        '''
+        Update all the signals with values from Agent.update().
+
+        Args
+
+        signal_values (list of float): Signals passed to Agent.update() by the Controller.
+        '''
+        for signal, value in zip(self._signals, signal_values):
+            signal.update(value)
+
+    def reset(self):
+        '''
+        Reset all the signals.
+        '''
+        for signal in self._signals:
+            signal.reset()
+
+    def append(self, signal):
+        '''
+        Appends a signal to the signal list
+
+        Args
+
+        signal (Signal): A Signal object.
+        '''
+        self._signals.append(signal)
+
+    def get_descriptors(self):
+        '''
+        Returns
+
+        A list of signal.get_descriptor() outputs which can be used in Agent.get_signals().
+        '''
+        return [sig.get_descriptor() for sig in self._signals]
+
+    def __getitem__(self, idx):
+        '''
+        Finds and returns a signal in this list via its index, tuple representation or signature
+        string. Usage is through [] notation. Example: sig_list["name@board-0"].
+
+        Args
+
+        idx (int or tuple of (str, domain, int) or str): Index, tuple representation or signature
+        string.
+        '''
+        return_val = None
+        if isinstance(idx, int):
+            return_val = self._signals[idx]
+        elif isinstance(idx, tuple):
+            for sig in self._signals:
+                if sig.get_descriptor() == idx:
+                    return_val = sig
+                    break
+        elif isinstance(idx, str):
+            for sig in self._signals:
+                if sig.get_signature() == idx:
+                    return_val = sig
+                    break
+
+        if return_val is None:
+            raise ValueError("Not found: " + str(idx))
+
+        return return_val
+
+
+class MetricList:
```

Comment:
It's not clear to me why this class is necessary.  If it's just a matter of tracking time, can't you just add it to the SignalList?

The reporting and tracing capabilities should me moved to a different class.